labcontrol/db/sheet.py

```diff
@@ -0,0 +1,1581 @@
+        x : str
+
+        Returns
+        -------
+        str
+            the input string with whitespaces replaced with underscores and
+            any other non-alphanumeric, non-hyphen, non-underscore characters
+            replaced with a hyphen.
+        """
+
+        # Replace any whitespace(s) with underscore
+        x = re.sub(r"\s+", '_', x)
+
+        # Replace any other character that is not alphanumeric, an underscore,
+        # or a hyphen (and thus valid in a folder name) with a hyphen
+        x = re.sub('[^0-9a-zA-Z-_]+', '-', x)
+        return x
+
+    @staticmethod
+    def _bcl_scrub_name(name):
+        """Modifies a sample name to be BCL2fastq compatible
+
+        Parameters
+        ----------
+        name : str
+            the sample name
+
+        Returns
+        -------
+        str
+            the sample name, formatted for bcl2fastq
+        """
+        return re.sub('[^0-9a-zA-Z-_]+', '_', name)
+
+    @staticmethod
+    def _set_control_values_to_plate_value(input_df, plate_col_name,
+                                           projname_col_name):
+        """ Update project name for control samples
+
+        Ensure that each sample plate included in the dataframe does not
+        contain experimental samples with more than (or less than) one
+        value in the column named projname_col_name. Assuming this is true, set
+        the project column value for each non-experimental sample to the value
+        of the project name for the (single) project on the non-experimental
+        sample's plate.
+
+        Parameters
+        ----------
+        input_df: pandas.DataFrame
+            A dataframe containing (at least) a column of plate names (having
+            the column name given in plate_col_name) and a column of project
+            names (having the column name given in projname_col_name)--e.g.,
+            Project Name on prep sheet or sample_proj_name on sample sheet--
+            and one row for each sample (both experimental and non-
+            experimental).  The value in the project name column must be None
+            for control (blank/positive control/etc) samples.
+        plate_col_name: str
+            The name of the column in input_df that contains the name of the
+            plate on which a given sample lies.
+        projname_col_name: str
+            The name of the column in input_df that contains the name of the
+            project associated with the given sample.
+
+        Returns
+        -------
+        result_df: pandas.DataFrame
+            A copy of the input dataframe, modified so that the controls have
+            the same (single) project name as the experimental samples on their
+            sample plate.
+
+        Raises
+        ------
+        ValueError
+            If any plate contains experimental samples from more (or fewer)
+            than one project.
+        """
+
+        assert plate_col_name in input_df.columns.values
```

Comment:
From what I understand, using `assert` statements in production code (i.e. not test code) is a bad idea since optimizers will remove these statements.

A safer and roughly equivalent way to write these two statements would be something like
```python
if plate_col_name not in input_df.columns.values:
    raise ValueError(
        "Specified plate column is not in the input DataFrame's columns"
    )
if projname_col_name not in input_df.columns.values:
    raise ValueError(
        "Specified project column is not in the input DataFrame's columns"
    )
```

Reply:
A good point, however I'm going to assume that the choice to use assert was intentional and the author was cognizant of that fact.